Add bookmarks

omaralbeik opened the pull request:
Hi guys 👋

We're trying to add bookmarking capability to FolioReaderKit

Here is what I did so far
1. Add a new `allowBookmarking` and `localizedBookmarkTitle` to `FolioReaderConfig`
2. Add new bookmark button to the navigation bar
3. Created a new `Bookmark` model (copied most of the code from the `Highlight` model) 😅 
4. Create a new `BookmarkList` UItableViewController
5. Add a new Bookmarks segment next to Contents and Highlights

What I was not able to do:
1. Get the current fragment visible on screen so I can forward the user back when tapping on a bookmark in the `BookmarkList` view controller
2. Set the navigation button's `isHighlighted` to `true` or `false` based on whether the currently visible page was bookmarked or not

@hebertialmeida your help is highly appreciated, it would be nice to see Bookmark capability in FolioReaderKit soon 💯

Commit:
Add bookmarks

Commit:
Remove commented code

omaralbeik commented:
cc @soyhanbeyazit

omaralbeik commented:
Any update on this @hebertialmeida :)

hebertialmeida commented:
Hey @omaralbeik, I haven't time to review that yet, but I will soon, I am on vacation right now... Sorry.

rodrigorsa commented on Source/FolioReaderBookmarkList.swift:
```diff
@@ -0,0 +1,95 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        self.tableView.register(UITableViewCell.self, forCellReuseIdentifier: kReuseCellIdentifier)
```
Could you remove all unnecessary `self`?

rodrigorsa commented on Source/FolioReaderBookmarkList.swift:
```diff
@@ -0,0 +1,95 @@
+        super.viewDidLoad()
+
+        self.tableView.register(UITableViewCell.self, forCellReuseIdentifier: kReuseCellIdentifier)
+        self.tableView.separatorInset = UIEdgeInsets.zero
```
Could you remove `UIEdgeInsets`?
`tableView.separatorInset = .zero`

rodrigorsa commented on Source/FolioReaderBookmarkList.swift:
```diff
@@ -0,0 +1,95 @@
+
+    // MARK: - Table view data source
+
+    override func numberOfSections(in tableView: UITableView) -> Int {
```
I think you could remove this method

rodrigorsa commented on Source/FolioReaderBookmarkList.swift:
```diff
@@ -0,0 +1,95 @@
+
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: kReuseCellIdentifier, for: indexPath)
+        cell.backgroundColor = UIColor.clear
```
Could you remove `UIColor`?
`cell.backgroundColor = .clear`

rodrigorsa commented on Source/FolioReaderBookmarkList.swift:
```diff
@@ -0,0 +1,95 @@
+        let cell = tableView.dequeueReusableCell(withIdentifier: kReuseCellIdentifier, for: indexPath)
+        cell.backgroundColor = UIColor.clear
+
+        let bookmark = bookmarks[(indexPath as NSIndexPath).row]
```
Why `(indexPath as NSIndexPath)`?

rodrigorsa commented on Source/FolioReaderBookmarkList.swift:
```diff
@@ -0,0 +1,95 @@
+        let bookmark = bookmarks[(indexPath as NSIndexPath).row]
+
+        // Format date
+        let dateFormatter = DateFormatter()
```
Could you turn it into a lazy var?
```
private lazy var dateFormatter: DateFormatter = {
   let dateFormatter = DateFormatter()
   dateFormatter.dateFormat = self.readerConfig.localizedHighlightsDateFormat
   return dateFormatter
}()
```

rodrigorsa commented on Source/FolioReaderBookmarkList.swift:
```diff
@@ -0,0 +1,95 @@
+    // MARK: - Table view delegate
+
+    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        guard let bookmark = bookmarks[safe: (indexPath as NSIndexPath).row] else {
```
Is `(indexPath as NSIndexPath)` really necessary?

rodrigorsa commented on Source/FolioReaderBookmarkList.swift:
```diff
@@ -0,0 +1,95 @@
+
+    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
+        if editingStyle == .delete {
+            let bookmark = bookmarks[(indexPath as NSIndexPath).row]
```
Here again..

rodrigorsa commented on Source/FolioReaderCenter.swift:
```diff
@@ -1054,53 +1060,68 @@ open class FolioReaderCenter: UIViewController, UICollectionViewDelegate, UIColl
     }
+    
+    @objc func didTapBookmarkIcon(_ sender: UIBookmarkBarButtonItem) {
+//        guard let bookId = (self.book.name as NSString?)?.deletingPathExtension else { return }
```
Is it leftover?

rodrigorsa replied:
If not, you could do `sender.isHighlighed = !sender.isHighlighed`

rodrigorsa commented on Source/Models/Bookmark+Helper.swift:
```diff
@@ -0,0 +1,130 @@
+                realm.delete(self)
+                try realm.commitWrite()
+            }
+        } catch let error as NSError {
```
You could remove `let error as NSError`

rodrigorsa commented on Source/Models/Bookmark+Helper.swift:
```diff
@@ -0,0 +1,130 @@
+            let realm = try Realm(configuration: readerConfig.realmConfiguration)
+            bookmark = realm.objects(Bookmark.self).filter(predicate).toArray(Bookmark.self).first
+            bookmark?.remove(withConfiguration: readerConfig)
+        } catch let error as NSError {
```
Also here

rodrigorsa commented on Source/Models/Bookmark+Helper.swift:
```diff
@@ -0,0 +1,130 @@
+            let realm = try Realm(configuration: readerConfig.realmConfiguration)
+            bookmarks = realm.objects(Bookmark.self).filter(predicate).toArray(Bookmark.self)
+            return (bookmarks ?? [])
+        } catch let error as NSError {
```
here too `let error as NSError`

rodrigorsa commented on Source/Models/Bookmark+Helper.swift:
```diff
@@ -0,0 +1,130 @@
+        do {
+            let realm = try Realm(configuration: readerConfig.realmConfiguration)
+            bookmarks = realm.objects(Bookmark.self).filter(predicate).toArray(Bookmark.self)
+            return (bookmarks ?? [])
```
Extra parenthesis `(...)`

rodrigorsa commented on Source/Models/Bookmark+Helper.swift:
```diff
@@ -0,0 +1,130 @@
+        do {
+            let realm = try Realm(configuration: readerConfig.realmConfiguration)
+            bookmarks = realm.objects(Bookmark.self).toArray(Bookmark.self)
+            return (bookmarks ?? [])
```
Extra parenthesis `(...)`

rodrigorsa commented on Source/Models/Bookmark+Helper.swift:
```diff
@@ -0,0 +1,130 @@
+            let realm = try Realm(configuration: readerConfig.realmConfiguration)
+            bookmarks = realm.objects(Bookmark.self).toArray(Bookmark.self)
+            return (bookmarks ?? [])
+        } catch let error as NSError {
```
here too `let error as NSError`

rodrigorsa commented on Source/Models/Bookmark.swift:
```diff
@@ -0,0 +1,21 @@
+
+/// A Bookmark object
+open class Bookmark: Object {
+    @objc open dynamic var bookId: String!
```
Why are you using `String!`? Could it be optional? If not, you shouldn't use `!`

Commit:
Update Bookmarks

omaralbeik commented:
Hey @rodrigorsa, thank you for your review, I updated the submission with fixes to some points you mentioned above.
- removed extra `self`s, I'm not a fan of stating it everywhere as well, I added it to follow the same coding style of the project - look at FolioReaderHighlightList :)
- as of the NSError, it is required because the `Completion` object expects `NSError`

I'd recommend adding [SwiftLint](https://github.com/realm/SwiftLint) to catch this kind of warnings, however, I'm afraid there are tons of changes required as the current code base follows a different code style

hebertialmeida committed:
Merge branch 'master' into add-bookmark

hebertialmeida commented:
This PR is really good, the only problem that I have not merged yet is the same problem with #334. 

Basically, we need to solve locator in the epub, that is the biggest problem we have right now, most of the bugs are highlight related and that being said anything we add will just increase bug reports.

Do you have experience with epub locators like [CFI](http://www.idpf.org/epub/linking/cfi/epub-cfi.html)? So you might help us solve this problem before merging that.

If we merge and people start using it we will have to probably implement some migration solution because the highlights format probably will change.

shalinipd commented:
@hebertialmeida @rodrigorsa Is this is being merged in the Repository as We are also looking for this (Bookmark) feature.
I have one question - Do Book search feature available in the Repository.

shubhankarbhavsar commented:
Hi @hebertialmeida ,

This library is good but has some bugs related to highlights. Especially when trying to highlight multiple paragraphs at once. 

@omaralbeik  Have you completed the bookmark feature? I require this feature for one of my projects. Still working on my own implementation for it.

shalinipd commented:
Anyone is ware that library have Book Search feature or not?

ynpl commented:
When can this library implement the bookmark function?

omaralbeik commented:
Sorry, I can't find the time to work on this feature anymore 😅

farhaansari commented:
Hi @omaralbeik,

Can you please guide me to achieve this? Can you please indicate in which direction I can look to make it work. Any help is highly appreciated.

shalinipd commented:
Hi @hebertialmeida : We are facing issue while highlighting multiple paragraphs at once. Can you please provide solution for the same.if anyone else has resolved it please provide the solution.

shubhankarbhavsar commented:
Hi @shalinipd : I had faced the same issue. I end up integrating rangy library in iOS to resolve this and that gives me an advantage as both platforms iOS and Android now have the same approach for Highlight.

shalinipd commented:
Hey ..thanks for replying.Can you give us the link of library as we are
searching the same for iOS and have not found.

On Wed, Jul 17, 2019 at 12:49 PM shubhankarbhavsar <notifications@github.com>
wrote:

> Hi @shalinipd <https://github.com/shalinipd> : I had faced the same
> issue. I end up integrating rangy library in iOS to resolve this and that
> gives me an advantage as both platforms iOS and Android now have the same
> approach for Highlight.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/FolioReader/FolioReaderKit/pull/328?email_source=notifications&email_token=ABWPL7646R2DTZHUXZ5C7OTP73BY7A5CNFSM4EXL7FF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2DI62A#issuecomment-512135016>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABWPL7ZIVUX4FKKEZ2KTSX3P73BY7ANCNFSM4EXL7FFQ>
> .
>

shalinipd commented:
@shubhankarbhavsar  If you can share the code base that will be helpful.

shubhankarbhavsar commented:
@shalinipd Unfortunately I do not have the code but I can help and guide.

shubhankarbhavsar commented:
> Hey ..thanks for replying.Can you give us the link of library as we are searching the same for iOS and have not found.
> […](#)
> On Wed, Jul 17, 2019 at 12:49 PM shubhankarbhavsar ***@***.***> wrote: Hi @shalinipd <https://github.com/shalinipd> : I had faced the same issue. I end up integrating rangy library in iOS to resolve this and that gives me an advantage as both platforms iOS and Android now have the same approach for Highlight. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <#328?email_source=notifications&email_token=ABWPL7646R2DTZHUXZ5C7OTP73BY7A5CNFSM4EXL7FF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2DI62A#issuecomment-512135016>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ABWPL7ZIVUX4FKKEZ2KTSX3P73BY7ANCNFSM4EXL7FFQ> .

Please use rangy js files from FolioReader Android project.

shubhankarbhavsar commented:
@shalinipd Sorry I don't have the code but I can help.

shalinipd commented:
Hi shubhankar,

As you suggested we are trying to integrate rangy in iOS but we are getting
error "undefined error".can you guide us how to integrate the same.

On Wed, Jul 17, 2019 at 3:44 PM shubhankarbhavsar <notifications@github.com>
wrote:

> @shalinipd <https://github.com/shalinipd> Unfortunately I do not have the
> code but I can help and guide.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/FolioReader/FolioReaderKit/pull/328?email_source=notifications&email_token=ABWPL72YFJP3PN6CMXCHYOTP73WG7A5CNFSM4EXL7FF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2DXFVY#issuecomment-512193239>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABWPL7ZH3VKOJVXMGLKXNZLP73WG7ANCNFSM4EXL7FFQ>
> .
>

shalinipd commented:
Hi @hebertialmeida ,@shubhankarbhavsar  We were able to integrate the rangy lib and successfully see the highlight by using "highlighselection" method but how to persist it so that redirection can be achieved.Can you please guide us on the same.

shubhankarbhavsar commented:
@shalinipd You can follow a similar approach as Highlight works here. As you said you've already achieved the highlighting part which means there is only one step left which is to return parameters from the highlight function. Highlighter's highlightSelection() method returns the highlightRangy string and you can prepare a custom JSON object by adding this in the parameter list and return it from the highlight function. After that, you can save this JSON object in any database like realm or core data.

shalinipd commented:
Hi FolioReader/FolioReaderKit

Thanks for your response.yes we are getting the JSON object which have
multiple values.How do we use these values to again put the highlight on
the original HTML.Can you please share a code sample and methods with us.

Please find attached the sample JSON which we are getting attached.


"[{\"id\":1,\"characterRange\":{\"start\":2,\"end\":31},\"doc\":{\"location\":{\"href\":\"file:///Users/dmi/Library/Developer/CoreSimulator/Devices/FCEF81F6-EEE3-4B0C-A1F5-09E7ACBA7734/data/Containers/Data/Application/1B12B7D9-AD63-432D-9BCB-35777CB7730E/Documents/English_DSM5.epub/OEBPS/xhtml/\",\"protocol\":\"file:\",\"host\":\"\",\"hostname\":\"\",\"port\":\"\",\"pathname\":\"/Users/dmi/Library/Developer/CoreSimulator/Devices/FCEF81F6-EEE3-4B0C-A1F5-09E7ACBA7734/data/Containers/Data/Application/1B12B7D9-AD63-432D-9BCB-35777CB7730E/Documents/English_DSM5.epub/OEBPS/xhtml/\",\"search\":\"\",\"hash\":\"\",\"origin\":\"file://\",\"ancestorOrigins\":{}}},\"classApplier\":{\"className\":\"highlight\",\"cssClass\":\"highlight\",\"ignoreWhiteSpace\":true,\"normalize\":true,\"attrExceptions\":[],\"elementProperties\":{},\"elementSortedClassName\":\"highlight\",\"applyToAnyTagName\":false,\"tagNames\":[\"span\",\"a\"]},\"converter\":{\"type\":\"textContent\"},\"containerElementId\":null,\"applied\":true}]"

Looking for your support we are really struck.

On Wed, Aug 7, 2019 at 4:43 PM shubhankarbhavsar <notifications@github.com>
wrote:

> @shalinipd <https://github.com/shalinipd> You can follow a similar
> approach as Highlight works here. As you said you've already achieved the
> highlighting part which means there is only one step left which is to
> return parameters from the highlight function. Highlighter's
> highlightSelection() method returns the highlightRangy string and you can
> prepare a custom JSON object by adding this in the parameter list and
> return it from the highlight function. After that, you can save this JSON
> object in any database like realm or core data.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/FolioReader/FolioReaderKit/pull/328?email_source=notifications&email_token=ABWPL76KEEUYAAMS4GO24V3QDKU3ZA5CNFSM4EXL7FF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3YBWYQ#issuecomment-519052130>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABWPL755KE43QT3CLYJ6Y3DQDKU3ZANCNFSM4EXL7FFQ>
> .
>

shubhankarbhavsar commented:
@shalinipd 
This is what I'm explaining in the last message. 

var highlightRangy = this.highlighter.highlightSelection(color, null, highlightId);
var params = { content: rangeString, rangy: highlightRangy, color: color, id: highlightId };
 return JSON.stringify(params);

Expected Rangy string returned in highlightRangy param should be like this: 69$79$1$highlight-yellow$

Please note that you must use the rangy files that are attached in FolioReader Android Project.

shalinipd commented:
Hi Subhankar,

Thanks for your reply.This was really helpful.We were able to integrate the
library.The last issue which we are facing after integrating rangy is its
not working in case of table content.Have you done any fix for the same.

We have logged this issue in folio also -
https://github.com/FolioReader/FolioReader-Android/issues/390

This is happening in Android library also.looking for your answer on the
same.

Thanks
Shalini Pandey


On Wed, Aug 7, 2019 at 7:49 PM shubhankarbhavsar <notifications@github.com>
wrote:

> @shalinipd <https://github.com/shalinipd>
> This is what I'm explaining in the last message.
>
> var highlightRangy = this.highlighter.highlightSelection(color, null,
> highlightId);
> var params = { content: rangeString, rangy: highlightRangy, color: color,
> id: highlightId };
> return JSON.stringify(params);
>
> Expected Rangy string returned in highlightRangy param should be like
> this: 69$79$1$highlight-yellow$
>
> Please note that you must use the rangy files that are attached in
> FolioReader Android Project.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/FolioReader/FolioReaderKit/pull/328?email_source=notifications&email_token=ABWPL7Y7FHTZTJ4OORENIQ3QDLKYRA5CNFSM4EXL7FF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3YR47Y#issuecomment-519118463>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABWPL7Z4H43Z7R7X5OFTH73QDLKYRANCNFSM4EXL7FFQ>
> .
>

shubhankarbhavsar commented:
Hi Shalini,

Congratulations on the success. I've not faced issue while highlighting table contents. I'll require code access to understand the issue. If possible please share something to look into.

shalinipd commented:
Hi Subhankar,

I tried with above epub file in Sample code provided by FolioReader Android
github page itself and found same issue. epub is attached in the mail.

Do you need any further information let me know.

Thanks
Shalini Pandey



On Mon, Aug 26, 2019 at 7:49 PM shubhankarbhavsar <notifications@github.com>
wrote:

> Hi Shalini,
>
> Congratulations on the success. I've not faced issue while highlighting
> table contents. I'll require code access to understand the issue. If
> possible please share something to look into.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/FolioReader/FolioReaderKit/pull/328?email_source=notifications&email_token=ABWPL7YRYT3CP2BUYCJT5LLQGPQ6BA5CNFSM4EXL7FF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5EQJ6Q#issuecomment-524879098>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABWPL73B4B6NK74PTUBAG5TQGPQ6BANCNFSM4EXL7FFQ>
> .
>

shalinipd commented:
Hi Subhankar,

really looking for your support.

Thanks
Shalini Pandey

On Tue, Aug 27, 2019 at 5:48 PM Shalini Pandey <shalini.pndy@gmail.com>
wrote:

> Hi Subhankar,
>
> I tried with above epub file in Sample code provided by FolioReader
> Android github page itself and found same issue. epub is attached in the
> mail.
>
> Do you need any further information let me know.
>
> Thanks
> Shalini Pandey
>
>
>
> On Mon, Aug 26, 2019 at 7:49 PM shubhankarbhavsar <
> notifications@github.com> wrote:
>
>> Hi Shalini,
>>
>> Congratulations on the success. I've not faced issue while highlighting
>> table contents. I'll require code access to understand the issue. If
>> possible please share something to look into.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/FolioReader/FolioReaderKit/pull/328?email_source=notifications&email_token=ABWPL7YRYT3CP2BUYCJT5LLQGPQ6BA5CNFSM4EXL7FF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5EQJ6Q#issuecomment-524879098>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABWPL73B4B6NK74PTUBAG5TQGPQ6BANCNFSM4EXL7FFQ>
>> .
>>
>
